### ISSUE TITLE
How to connect protected wifi

### ISSUE THREAD

**Issue by maffo01** [sidorares/dbus-native#297]
I tried to write js code to connect to a secure network with dbus-native for connman. Has anyone tried this yet? I think I can't seem to create the RequestInput and get it to work after a connect

**Comment by sidorares** [comment]
Answerable now — and the piece that was missing is that an agent is a service *you* export, which ConnMan then calls back into.

`RequestInput` is not something you call. ConnMan calls it **on you**: you register an object implementing `net.connman.Agent`, and when a network needs a passphrase ConnMan invokes `RequestInput` on that object and waits for your reply. So the order is register → export → wait to be called, and the connect only completes once you have answered.

```js
const dbus = require('dbus-native');
const { defineInterface } = dbus;

const bus = dbus.systemBus();

const agent = defineInterface({
  name: 'net.connman.Agent',
  methods: {
    // (o, a{sv}) -> (a{sv}) : ConnMan says what it wants, you answer
    RequestInput: {
      in: { service: 'o', fields: 'a{sv}' },
      out: { reply: 'a{sv}' },
      handler: ({ service, fields }) => {
        // `fields` names what is being asked for -- Passphrase, Name, ...
        return { Passphrase: 'my-wifi-password' };
      }
    },
    Release: { in: {}, out: {}, handler: () => {} },
    Cancel: { in: {}, out: {}, handler: () => {} },
    ReportError: { in: { service: 'o', error: 's' }, out: {}, handler: ({ service, error }) => {} }
  }
});

await using reg = await bus.export('/net/connman/Agent', agent);

// tell ConnMan the agent exists -- before connecting
await bus.invoke({
  destination: 'net.connman',
  path: '/',
  interface: 'net.connman.Manager',
  member: 'RegisterAgent',
  signature: 'o',
  body: ['/net/connman/Agent']
});

// this can now call back into RequestInput
const svc = await bus.proxy('net.connman', '/net/connman/service/wifi_...');
await svc.Connect();
```

Note the handler shape: `in` and `out` are objects of **named** arguments, and a handler receives one object of those names (plus a context object with `sender`, `path`, and so on). With a single `out` you return the value itself, which is why `RequestInput` returns the dict directly.

Three things that are easy to get wrong, and are the likely reason it "didn't work after connect":

- **Register the agent before calling `Connect()`.** With no agent registered when input is needed, ConnMan fails the connect rather than waiting.
- **The reply is `a{sv}`, and a plain object is all it needs.** Writing into an `a{sv}` infers each value's signature, so `{ Passphrase: 'secret' }` goes out as `s` — identical bytes to `new Variant('s', 'secret')`. Every field ConnMan asks for here is a string, so no `Variant` is needed anywhere in this call. Reach for `Variant` only to override the guess or to reach a type inference cannot produce (`u`, `y`, `o`, `g`, structs, `av`) — a plain `3` is sent as `i`.
- **`Connect()` can take a long time** — ConnMan waits on association and DHCP. Since 0.14.0 calls have a 25-second default deadline, so pass `{ timeout: 60000 }` (or `0`) for that one call, or it will reject while ConnMan is still working.

`defineInterface()` and `bus.export()` shipped in 0.13.0, and the disposable registration means the agent unregisters when the scope ends.

Closing as answered. I have not tested this against ConnMan specifically — the export-and-be-called-back pattern is exercised against a real daemon in the integration suite, but the ConnMan-specific details above come from its API docs rather than from a running system. If it does not work, please open a new issue with what ConnMan reports.

---

*Edited 2026-07-30, twice. The original said the `a{sv}` values had to be wrapped in `Variant` — they do not; a plain object is enough, and for a string it produces identical bytes (measurements in the follow-up comment). It also wrote `in`/`out` as arrays with positional handler arguments, which is not the contract: they are objects of named arguments and a handler receives one object. Both corrected above, and the example has now been run against a real daemon.*

**Comment by sidorares** [comment]
**Correction to the comment above:** the `Variant` wrapper is not needed there. `{ Passphrase: 'my-wifi-password' }` is enough, and my "not a bare string" was wrong.

Writing a plain object into an `a{sv}` infers each value's signature. Checked against a real daemon, echoing back the types that arrived on the wire:

```
{ Passphrase: 'secret' }                 ->  Passphrase:s="secret"
{ Passphrase: new Variant('s','secret') }->  Passphrase:s="secret"
{ N: 42 }                                ->  N:i=42
{ B: true }                              ->  B:b=true
{ A: ['x','y'] }                         ->  A:as=["x","y"]
{ S: 'a', N: 7, B: false }               ->  S:s="a"  N:i=7  B:b=false
```

For a string the two forms produce identical bytes. So the handler is simply:

```js
RequestInput: {
  in: ['o', 'a{sv}'],
  out: ['a{sv}'],
  handler: (servicePath, fields) => ({ Passphrase: 'my-wifi-password' })
}
```

and all of ConnMan's `RequestInput` reply fields — `Passphrase`, `Name`, `Identity`, `Username`, `Password` — are strings, so nothing in that call needs a `Variant` at all.

**When you do need one.** Inference cannot produce every type, and for numbers it has to pick:

| JavaScript | inferred |
| --- | --- |
| `string` | `s` |
| `boolean` | `b` |
| integer within int32 | `i` |
| integer outside int32 | `x` |
| any other number | `d` |
| `Buffer` | `ay` |
| homogeneous array | `a` + element type |
| plain object | `a{sv}` |

So a plain `3` is `i`, and if the peer wants `u`, `y`, `o`, `g`, a struct or `av`, you have to say so:

```js
{ Count: new Variant('u', 3) }   // uint32, not int32
```

That is the rule: **`Variant` is for overriding the guess or reaching a type inference cannot produce — not a requirement of `a{sv}` itself.** Full table in [docs/api.md](https://github.com/sidorares/dbus-native/blob/master/docs/api.md#values-and-types).

Thanks for the catch — the original comment made this look more ceremonious than it is.
